File: TradeItIosEmsApi/TradeItConnector.m
//
//  TradeItConnector.m
//  TradeItIosEmsApi
//
//  Created by Antonio Reyes on 1/12/16.
//  Copyright © 2016 TradeIt. All rights reserved.
//

#import "TradeItConnector.h"
#import "TradeItKeychain.h"

#ifdef CARTHAGE
#import <TradeItIosTicketSDK2Carthage/TradeItIosTicketSDK2Carthage-Swift.h>
#else
    #if __has_include("TradeItIosTicketSDK2-Swift.h")
        #import "TradeItIosTicketSDK2-Swift.h"
    #else
        #import <TradeItIosTicketSDK2/TradeItIosTicketSDK2-Swift.h>
    #endif
#endif

@interface TradeItConnector()

- (NSUserDefaults *)userDefaults;

@end

@implementation TradeItConnector {
    BOOL runAsyncCompletionBlockOnMainThread;
}

NSString *BROKER_LIST_KEYNAME = @"TRADEIT_BROKERS";
NSString *USER_DEFAULTS_SUITE = @"TRADEIT";

- (NSUserDefaults *)userDefaults {
    static NSUserDefaults *userDefaults = nil;
    static dispatch_once_t onceToken;

    dispatch_once(&onceToken, ^{
        userDefaults = [[NSUserDefaults alloc] initWithSuiteName:USER_DEFAULTS_SUITE];
    });

    return userDefaults;
}

- (id)initWithApiKey:(NSString *)apiKey
         environment:(TradeitEmsEnvironments)environment
             version:(TradeItEmsApiVersion)version {
    self = [super init];

    if (self) {
        self.apiKey = apiKey;
        self.environment = environment;
        self.version = version;
        runAsyncCompletionBlockOnMainThread = true;
        [self initSession];
    }

    return self;
}

- (id)initWithApiKey:(NSString *)apiKey {
    self = [super init];

    if (self) {
        self.apiKey = apiKey;
        self.environment = TradeItEmsProductionEnv;
        self.version = TradeItEmsApiVersion_2;
        runAsyncCompletionBlockOnMainThread = true;
        [self initSession];
    }

    return self;
}

- (TradeItLinkedLogin *)updateKeychainWithLink:(TradeItAuthLinkResult *)link
                                    withBroker:(NSString *)broker
                            withBrokerLongName:(NSString *)brokerLongName {
    NSDictionary *linkDict = [self getLinkedLoginDictByuserId:link.userId];

    if (linkDict) {
        // If the saved link is found, update the token in the keychain for its keychainId
        NSString *keychainId = linkDict[@"keychainId"];

        [TradeItKeychain saveString:link.userToken forKey:keychainId];

        return [[TradeItLinkedLogin alloc] initWithLabel:linkDict[@"label"]
                                                  broker:broker
                                                  brokerLongName:brokerLongName
                                                  userId:link.userId
                                              keyChainId:keychainId];
    } else {
        // No existing link for that userId so make a new one
        TradeItAuthLinkResult *authLinkResult = [[TradeItAuthLinkResult alloc] init];
        authLinkResult.userId = link.userId;
        authLinkResult.userToken = link.userToken;

        return [self saveToKeychainWithLink:authLinkResult
                                 withBroker:broker
                         withBrokerLongName:brokerLongName];
    }
}

- (TradeItLinkedLogin *)saveToKeychainWithLink:(TradeItAuthLinkResult *)link
                                    withBroker:(NSString *)broker
                                    withBrokerLongName:(NSString *)brokerLongName {
    return [self saveToKeychainWithLink:link withBroker:broker withBrokerLongName:brokerLongName andLabel:broker];
}

- (TradeItLinkedLogin *)saveToKeychainWithLink:(TradeItAuthLinkResult *)link
                                    withBroker:(NSString *)broker
                            withBrokerLongName:(NSString *)brokerLongName
                                      andLabel:(NSString *)label {
    return [self saveToKeychainWithUserId:link.userId andUserToken:link.userToken andBroker:broker andBrokerLongName:brokerLongName andLabel:label];
}

- (TradeItLinkedLogin *)saveToKeychainWithUserId:(NSString *)userId
                                    andUserToken:(NSString *)userToken
                                       andBroker:(NSString *)broker
                               andBrokerLongName:(NSString *)brokerLongName
                                        andLabel:(NSString *)label {
    NSMutableArray *accounts = [[NSMutableArray alloc] initWithArray:[self getLinkedLoginsRaw]];
    NSString *keychainId = [[NSUUID UUID] UUIDString];

    NSDictionary *newRecord = @{@"label":label,
                                @"broker":broker,
                                @"brokerLongName":brokerLongName,
                                @"userId":userId,
                                @"keychainId":keychainId};

    [accounts addObject:newRecord];

    [self.userDefaults setObject:accounts forKey:BROKER_LIST_KEYNAME];

    [TradeItKeychain saveString:userToken forKey:keychainId];

    return [[TradeItLinkedLogin alloc] initWithLabel:label
                                              broker:broker
                                      brokerLongName: brokerLongName
                                              userId:userId
                                          keyChainId:keychainId];
}

- (NSDictionary *)getLinkedLoginDictByuserId:(NSString *)userId {
    NSArray *linkedLoginDicts = [self getLinkedLoginsRaw];

    // Search for the existing saved link by userId
    NSPredicate *filter = [NSPredicate predicateWithBlock:^BOOL(NSDictionary *linkDict, NSDictionary * __unused bindings) {
        return [linkDict[@"userId"] isEqual:userId];
    }];

    NSArray *filteredLinkDicts = [linkedLoginDicts filteredArrayUsingPredicate:filter];

    if (filteredLinkDicts.count > 0) {
        // Link found
        return filteredLinkDicts[0];
    } else {
        // Link not found
        return nil;
    }
}

- (NSArray *)getLinkedLoginsRaw {
    NSArray *linkedAccounts = [self.userDefaults arrayForKey:BROKER_LIST_KEYNAME];

    if (!linkedAccounts) {
        linkedAccounts = [[NSArray alloc] init];
    }

    /*
     NSLog(@"------------Linked Logins-------------");
     [linkedAccounts enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
     NSDictionary * account = (NSDictionary *) obj;
     NSLog(@"Broker: %@ - Label: %@ - UserId: %@ - KeychainId: %@", account[@"broker"], account[@"label"], account[@"userId"], account[@"keychainId"]);
     }];
     */

    return linkedAccounts;
}

- (NSArray *)getLinkedLogins {
    NSArray *linkedAccounts = [self getLinkedLoginsRaw];

    NSMutableArray *accountsToReturn = [[NSMutableArray alloc] init];
    for (NSDictionary *account in linkedAccounts) {
        [accountsToReturn addObject:[[TradeItLinkedLogin alloc] initWithLabel:account[@"label"]
                                                                       broker:account[@"broker"]
                                                                       brokerLongName:account[@"brokerLongName"] ?: account[@"broker"]
                                                                       userId:account[@"userId"]
                                                                   keyChainId:account[@"keychainId"]]];
    }

    return accountsToReturn;
}

- (void)deleteLocalLinkedLogin:(TradeItLinkedLogin *)login {
    NSMutableArray *accounts = [[NSMutableArray alloc] initWithArray:[self getLinkedLoginsRaw]];
    NSMutableArray *toRemove = [[NSMutableArray alloc] init];

    [accounts enumerateObjectsUsingBlock:^(id _Nonnull obj, NSUInteger __unused idx, BOOL * _Nonnull __unused stop) {
        NSDictionary *account = (NSDictionary *)obj;
        if ([account[@"userId"] isEqualToString:login.userId]) {
            [toRemove addObject:obj];
        }
    }];

    for (NSDictionary *account in toRemove) {
        [accounts removeObject:account];
    }

    [self.userDefaults setObject:accounts forKey:BROKER_LIST_KEYNAME];
}

-(void) initSession {
    NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
    configuration.requestCachePolicy = NSURLRequestReloadIgnoringLocalCacheData;
    configuration.URLCache = nil;
    self.session = [NSURLSession sessionWithConfiguration:configuration];
}

@end

File: TradeItIosTicketSDK2/TradeItConnector.swift
import PromiseKit

class TradeItConnector: NSObject {
    var apiKey: String
    var environment: TradeitEmsEnvironments
    var version: TradeItEmsApiVersion
    var session: URLSession
    
    static let BROKER_LIST_KEYNAME = "TRADEIT_BROKERS"
    static let USER_DEFAULTS_SUITE = "TRADEIT"
    
    private static let userDefaultsVar: UserDefaults? = {
        var userDefaults = UserDefaults(suiteName: USER_DEFAULTS_SUITE)
        return userDefaults
    }()
    
    static func userDefaults() -> UserDefaults? {
        return TradeItConnector.userDefaultsVar
    }
    
    init(apiKey: String, environment: TradeitEmsEnvironments, version: TradeItEmsApiVersion) {
        self.apiKey = apiKey
        self.environment = environment
        self.version = version
        self.session = URLSession(configuration: .default) // Swift compiler requires to initialize session directly in the constructor
        super.init()
        self.initSession()
    }
    
    init(apiKey: String) {
        self.apiKey = apiKey
        self.environment = TradeitEmsEnvironments.tradeItEmsProductionEnv
        self.version = TradeItEmsApiVersion._2
        self.session = URLSession(configuration: .default) // Swift compiler requires to initialize session directly in the constructor
        super.init()
        self.initSession()
    }
    
    
    func updateKeychain(withLink link: TradeItAuthLinkResult?, withBroker broker: String?, withBrokerLongName brokerLongName: String?) -> TradeItLinkedLogin? {
        let linkDict = getLinkedLoginDictByuserId(link?.userId)
        
        if let linkDict = linkDict {
            // If the saved link is found, update the token in the keychain for its keychainId
            let keychainId = linkDict["keychainId"] as? String
            
            TradeItKeychain.save(link?.userToken, forKey: keychainId)
            let label = linkDict["label"] as? String
            return TradeItLinkedLogin(
                label: label ?? "",
                broker: broker ?? "",
                brokerLongName: brokerLongName ?? "",
                userId: link?.userId ?? "",
                keychainId: keychainId ?? ""
            )
        } else {
            // No existing link for that userId so make a new one
            let authLinkResult = TradeItAuthLinkResult()
            authLinkResult.userId = link?.userId ?? ""
            authLinkResult.userToken = link?.userToken ?? ""
            return saveToKeychain(
                withLink: authLinkResult,
                withBroker: broker,
                withBrokerLongName: brokerLongName
            )
        }
    }
    
    func saveToKeychain(withLink link: TradeItAuthLinkResult?, withBroker broker: String?, withBrokerLongName brokerLongName: String?) -> TradeItLinkedLogin? {
        return saveToKeychain(withLink: link, withBroker: broker, withBrokerLongName: brokerLongName, andLabel: broker)
    }
    
    func saveToKeychain(withLink link: TradeItAuthLinkResult?, withBroker broker: String?, withBrokerLongName brokerLongName: String?, andLabel label: String?) -> TradeItLinkedLogin? {
        return saveToKeychain(withUserId: link?.userId, andUserToken: link?.userToken, andBroker: broker, andBrokerLongName: brokerLongName, andLabel: label)
    }
    
    func saveToKeychain(withUserId userId: String?, andUserToken userToken: String?, andBroker broker: String?, andBrokerLongName brokerLongName: String?, andLabel label: String?) -> TradeItLinkedLogin? {
        var accounts = getLinkedLoginsRaw()
        let keychainId = UUID().uuidString
        
        let newRecord = [
            "label": label ?? "",
            "broker": broker ?? "",
            "brokerLongName": brokerLongName ?? "",
            "userId": userId ?? "",
            "keychainId": keychainId
        ]
        
        accounts.append(newRecord)
        
        TradeItConnector.userDefaults()?.set(accounts, forKey: TradeItConnector.BROKER_LIST_KEYNAME)
        
        TradeItKeychain.save(userToken, forKey: keychainId)
        
        return TradeItLinkedLogin(
            label: label ?? "",
            broker: broker ?? "",
            brokerLongName: brokerLongName ?? "",
            userId: userId ?? "",
            keychainId: keychainId
        )
    }
    
    func getLinkedLoginDictByuserId(_ userId: String?) -> [AnyHashable : Any]? {
        let linkedLoginDicts = getLinkedLoginsRaw()
        
        // Search for the existing saved link by userId
        let filteredLinkDicts = linkedLoginDicts.filter { $0["userId"] == userId }
        
        if filteredLinkDicts.count > 0 {
            // Link found
            return filteredLinkDicts.first
        } else {
            // Link not found
            return nil
        }
    }
    
    func getLinkedLoginsRaw() -> [[String: String]] {
        var linkedAccounts = TradeItConnector.userDefaultsVar?.array(forKey: TradeItConnector.BROKER_LIST_KEYNAME)
        
        if linkedAccounts == nil {
            linkedAccounts = [[String : String]]()
        }
        
        /*
         NSLog(@"------------Linked Logins-------------");
         [linkedAccounts enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
         NSDictionary * account = (NSDictionary *) obj;
         NSLog(@"Broker: %@ - Label: %@ - UserId: %@ - KeychainId: %@", account[@"broker"], account[@"label"], account[@"userId"], account[@"keychainId"]);
         }];
         */
        
        return linkedAccounts as? [[String: String]] ?? []
    }
    
    func getLinkedLogins() -> [Any]? {
        let linkedAccounts = getLinkedLoginsRaw()
        
        var accountsToReturn: [TradeItLinkedLogin] = []
        for account: [String : String] in linkedAccounts {
            accountsToReturn.append(
                TradeItLinkedLogin(
                    label: account["label"] ?? "",
                    broker: account["broker"] ?? "",
                    brokerLongName: account["brokerLongName"] ?? account["broker"] ?? "",
                    userId: account["userId"] ?? "",
                    keychainId: account["keychainId"] ?? ""
                )
            )
        }
        
        return accountsToReturn
    }
    
    func deleteLocalLinkedLogin(_ login: TradeItLinkedLogin?) {
        var accounts = getLinkedLoginsRaw()
        var toRemove: [[String: String]] = []
        
        (accounts as NSArray).enumerateObjects({ obj, idx, stop in
            let account = obj as? [String : String] ?? [:]
            if (account["userId"] == login?.userId) {
                toRemove.append(account)
            }
        })
        
        accounts.removeAll { toRemove.contains($0)}
        
        TradeItConnector.userDefaultsVar?.set(accounts, forKey: TradeItConnector.BROKER_LIST_KEYNAME)
    }
    
    func userToken(fromKeychainId keychainId: String) -> String? {
        return TradeItKeychain.getStringForKey(keychainId)
    }
    
    func sendReturnJSON(
        _ request: URLRequest,
        withCompletionBlock completionBlock: @escaping (TradeItResult, String?) -> Void
        ) {
        self.send(request, withCompletionBlock: completionBlock)
    }
    
    func send<T: TradeItResult>(
        _ request: URLRequest,
        targetClassType: T.Type,
        withCompletionBlock completionBlock: @escaping (TradeItResult?) -> Void
        ) {
        self.send(request) { result, json in
            if result.isSuccessful() { // Try to cast to desired resultClass
                completionBlock(TradeItResultTransformer.transform(targetClassType: targetClassType, json: json))
            } else {
                completionBlock(result) // Review order or Security question or Error case
            }
        }
    }
    
    func send<T: TradeItResult>(
        _ request: URLRequest,
        targetClassType: T.Type
        ) -> Promise<T> {
        return Promise<T> { seal in
            send(request, targetClassType: targetClassType) { result in
                switch(result) {
                case let result as T: seal.fulfill(result)
                case let error as TradeItErrorResult: seal.reject(error)
                default:
                    seal.reject(
                        TradeItErrorResult.error(withSystemMessage: "The server returned a response that could not deserialize to the requested type: \(targetClassType.classForCoder())")
                    )
                }
            }
        }
    }
    
    // MARK: Private
    
    private func send(
        _ request: URLRequest,
        withCompletionBlock completionBlock: @escaping (TradeItResult, String?) -> Void
        ) {
        guard !TradeItSDK.isDeviceJailbroken else {
            completionBlock(
                TradeItErrorResult(
                    title: "This device is jailbroken",
                    message: "This action is not allowed on a jailbroken device"
                ), ""
            )
            return
        }
        
        if TradeItSDK.debug {
            let requestBodyString = String(data: request.httpBody ?? Data(), encoding: String.Encoding.utf8)
            print("\n===== REQUEST =====\n\(request.url?.absoluteString ?? "NO URL!")\n\(requestBodyString ?? "NO BODY!")\n")
        }
        
        DispatchQueue.global(qos: .userInitiated).async {
            self.session.dataTask(
                with: request,
                completionHandler: { data, response, error in
                    if TradeItSDK.debug {
                        let responseBodyString = String(data: data ?? Data(), encoding: String.Encoding.utf8)
                        print("\n===== RESPONSE =====\n\(request.url?.absoluteString ?? "NO URL!")\n\(responseBodyString ?? "NO BODY!")\n")
                        if let error = error {
                            print("Error: \(error.localizedDescription)")
                        }
                    }
                    
                    let (result, json) = self.processResponse(data, response, error)
                    
                    DispatchQueue.main.async { completionBlock(result, json) }
            }
                ).resume()
        }
    }
    
    private func processResponse(_ data: Data?, _ response: URLResponse?, _ error: Error?) -> (TradeItResult, String?) {
        guard let httpResponse = response as? HTTPURLResponse else {
            return (TradeItErrorResult.error(withSystemMessage: "Unable to cast response to HTTPUrlResponse. Error description message: \(error?.localizedDescription ?? "nil")"), nil)
        }
        
        guard httpResponse.statusCode == 200 else {
            return (TradeItErrorResult.error(withSystemMessage: "Response status code: \(httpResponse.statusCode)."), nil)
        }
        
        guard let data = data, let json = String(data: data, encoding: .utf8) else {
            return (TradeItErrorResult.error(withSystemMessage: "Unable to read JSON data."), nil)
        }
        
        var result = TradeItResultTransformer.transform(targetClassType: TradeItResult.self, json: json)
        
        if result?.isError() == true { // Server sent an ERROR response so try create a TradeItErrorResult
            result = TradeItResultTransformer.transform(targetClassType: TradeItErrorResult.self, json: json)
        } else if result?.isSecurityQuestion() == true {
            result = TradeItResultTransformer.transform(targetClassType: TradeItSecurityQuestionResult.self, json: json)
        }
        
        let defaultedResult = result ?? TradeItErrorResult.error(withSystemMessage: "JSON from server does not match the TradeItResult format.")
        
        return (defaultedResult, json)
    }
    
    private func initSession() {
        let configuration = URLSessionConfiguration.default
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        configuration.urlCache = nil
        session = URLSession(configuration: configuration)
    }
}
